singularity-eos/eos/eos_powermg.hpp

```diff
@@ -0,0 +1,440 @@
+  static std::string EosType() { return std::string("PowerMG"); }
+  static std::string EosPyType() { return EosType(); }
+
+ private:
+  static constexpr const unsigned long _preferred_input =
+      thermalqs::density | thermalqs::specific_internal_energy;
+  Real _rho0, _T0, _G0, _Cv0, _E0, _S0, _Pmin;
+  static constexpr const int PressureCoeffsK0toK40Size = 41;
+  int _M;
+  Real _K0toK40[PressureCoeffsK0toK40Size];
+  void _CheckPowerMG();
+  void _InitializePowerMG(const Real *expcoeffs);
+  PORTABLE_INLINE_FUNCTION Real _HugPressureFromDensity(const Real rho) const;
+  PORTABLE_INLINE_FUNCTION Real _HugTemperatureFromDensity(const Real rho) const;
+  PORTABLE_INLINE_FUNCTION Real _SN2Mp2(const Real x) const;
+  PORTABLE_INLINE_FUNCTION Real
+  _compBulkModulusFromDensityTemperature(const Real rho, const Real temp) const;
+  PORTABLE_INLINE_FUNCTION Real
+  _compBulkModulusFromDensityInternalEnergy(const Real rho, const Real sie) const;
+};
+
+inline void PowerMG::_CheckPowerMG() {
+
+  if (_rho0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter rho0 < 0");
+  }
+  if (_T0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter T0 < 0");
+  }
+  if (_Cv0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter Cv0 < 0");
+  }
+  if (_K0toK40[0] < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter K0 < 0");
+  }
+  if (_Pmin >= 0.0) {
+    _Pmin = -1000 * _K0toK40[0];
+#ifndef NDEBUG
+    PORTABLE_WARN(
+        "PowerMG model parameter Pmin not set or positive. Reset to default (-1000*K0)");
+#endif // NDEBUG
+  }
+}
+
+inline void PowerMG::_InitializePowerMG(const Real *K0toK40input) {
+
+  // The PressureCoeffsK0toK40Size (=41) allowed K0 to K40 coefficients
+  // for the pressure reference curve vs rho
+  // are seldom all used so did not want to crowd the argument list with them.
+  // Instead I ask the host code to send me a pointer to this array so that I can
+  // copy it here. Not used coeffs should be set to 0.0 (of course).
+  for (int ind = 0; ind < PressureCoeffsK0toK40Size; ind++) {
+    _K0toK40[ind] = K0toK40input[ind];
+    if (_K0toK40[ind] != 0.0) _M = ind;
+  }
+}
+
+PORTABLE_INLINE_FUNCTION Real PowerMG::_HugPressureFromDensity(Real rho) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  Real value = 0.0;
+  for (int ind = _M; ind >= 1; ind--) {
+    value = eta * value + _K0toK40[ind];
+  }
+  value = _K0toK40[0] * eta * (eta * value + 1.0);
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::_HugTemperatureFromDensity(Real rho) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  Real gamma2 = _SN2Mp2(_G0 * eta);
+  Real sum = _M * _K0toK40[_M] * gamma2;
+  for (int ind = _M - 1; ind >= 1; ind--) {
+    gamma2 = _G0 * eta / (ind + 2) * gamma2 + 1.0 / (ind + 2);
+    sum = eta * sum + ind * _K0toK40[ind] * gamma2;
+  }
+  Real temp = _K0toK40[0] / _Cv0 / 2.0 / _rho0 * eta * eta * eta * sum;
```

Comment:
Another minor clarity thing... Since `eta*eta*eta*sum` presumably is in the numerator here, I think it would help to put it before the divisions.

Reply:
Agree. Why I did it like this is because _K0toK40[0] / _Cv0 / 2.0 / _rho0 is a common refactor that shows up other places.

Reply:
refactor = prefactor

Reply:
That makes sense. Maybe put the prefactor in parentheses? That should make it clear to a casual reader what's going on.

Reply:
Done